File: app/src/main/java/com/bubelov/coins/repository/currency/CurrenciesRepository.java
package com.bubelov.coins.repository.currency;

import android.support.annotation.NonNull;

import com.bubelov.coins.model.Currency;

import java.io.IOException;
import java.util.Collection;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * @author Igor Bubelov
 */

@Singleton
public class CurrenciesRepository {
    private CurrenciesDataSourceNetwork networkDataSource;

    private CurrenciesDataSourceDisk diskDataSource;

    @Inject
    CurrenciesRepository(CurrenciesDataSourceNetwork networkDataSource, CurrenciesDataSourceDisk diskDataSource) {
        this.networkDataSource = networkDataSource;
        this.diskDataSource = diskDataSource;
    }

    public Currency getCurrency(@NonNull String code) {
        return diskDataSource.getCurrency(code);
    }

    public void reloadFromApi() throws IOException {
        Collection<Currency> currencies = networkDataSource.getCurrencies();
        diskDataSource.insert(currencies);
    }
}

File: app/src/main/java/com/bubelov/coins/repository/currency/CurrenciesRepository.kt
package com.bubelov.coins.repository.currency

import com.bubelov.coins.model.Currency

import java.io.IOException

import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author Igor Bubelov
 */

@Singleton
class CurrenciesRepository @Inject
internal constructor(private val apiDataSource: CurrenciesDataSourceApi, private val dbDataSource: CurrenciesDataSourceDb) {
    fun getCurrency(code: String): Currency? {
        return dbDataSource.getCurrency(code)
    }

    @Throws(IOException::class)
    fun reloadFromApi() {
        dbDataSource.insertCurrencies(apiDataSource.getCurrencies())
    }
}